doc/gold.rst

```diff
@@ -0,0 +1,412 @@
+the following groups:
+
+- whitespace, any of SPACE, TAB, LINE FEED, or CARRIAGE RETURN
+
+- special delimiters, any of ``( ) [ ] { } : ; . , " \ @ &``
+
+- alphanumeric characters, any of ``A-Za-z0-9``. Note that the underscore is not
+  included (though it is allowed in identifiers, but discouraged).
+
+- punctuation, any of ``! $ % ' * + - / < = > ? ^ _ ` | ~`` (i.e. the remaining
+  ASCII characters).
+
+Gold-syntax tokens are delimited by whitespace, which is discarded, and the
+special delimiters, which are considered single tokens. Note that new line
+characters are not treated specially. Also note that comments are treated
+specially by the parser; see below.
+
+A token beginning with a digit in ``0-9`` is assumed to be a numeric value. A
+valid numeric value must be one of the following:
+
+1. A sequence of characters in ``0-9``, such as ``1``, ``42``, or ``094``,
+   denotes a decimal (base-10) integer. Note that numbers beginning with a ``0``
+   are considered to be decimal, not octal.
+
+2. The characters ``0x`` or ``0X`` (a digit zero followed by an uppercase or
+   lowercase letter 'X') followed by a sequence of characters in ``0-9A-Fa-f``,
+   such as ``0x1``, ``0xa9``, or ``0xbeef``, denotes a hexadecimal (base-16)
+   integer.
+
+3. The characters ``0o`` or ``0O`` (a digit zero followed by an uppercase or
+   lowercase letter 'O') followed by a sequence of characters in ``0-7``, such
+   as ``0o644``, denotes an octal (base-8) integer. Note that numbers beginning
+   with a ``0`` (zero) followed by digits are considered to be decimal, not
+   octal.
+
+4. The characters ``0b`` or ``0B`` (a digit zero followed by an uppercase or
+   lowercase letter 'B') followed by a sequence of characters in ``0 1``, such
+   as ``0b11010001``, denotes a binary (base-2) integer.
+
+A token beginning with a digit in ``0-9`` and not conforming to one of the above
+rules is an error.
+
+A token beginning with a letter in ``A-Za-z`` is assumed to be an identifier.
+The restriction to a letter applies only to the first character; identifiers may
+contain any printable ASCII character which does not delimit a token. For
+example, ``foo-bar?`` is considered a valid identifier.
+
+A token beginning with a punctuation character continues until the first
+non-punctuation character, at which point it will end, regardless of whether a
+whitespace character, comment, or special delimiter has been encountered.
+Therefore ``-1`` becomes two tokens.
+
+.. attention:: The above rules imply that the string ``1+2`` is considered a
+               single, invalid, numeric token, and the string ``x-2`` is
+               considered an identifier. Make sure to put spaces around your
+               operators!
+
+
+Language Elements
+=================
+
+A Gold-syntax unit consists of a series of top-level statements.
+
+Binding Statements
+------------------
+
+The following statements introduce bindings of names. They are all allowed at
```

Comment:
"bindings of names" -> let's say "bindings of identifiers" to be consistent with other verbiage we've used.

let's standardize on either "names" or "identifiers" throughout this document. unless there's a nuance i'm overlooking?

Reply:
I think there was meant to be a nuance, but if so, I should have called it out. I think it was something like "identifier" refers to the token itself, while things can have identifiers as names.

If you think this is drawing shades of meaning too fine, I wouldn't mind standardizing on "name" (it flows better in English prose), though I think there might be prior art for this distinction.